addons/controllers/csi/vspherecsiconfig_controller.go

```diff
@@ -0,0 +1,173 @@
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=vspherecsiconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=vspherecsiconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.tanzu.vmware.com,resources=vspherecsiconfigs/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *VSphereCSIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("VSphereCSIConfig", req.NamespacedName)
+
+	vcsiConfig := &csiv1alpha1.VSphereCSIConfig{}
+	if err := r.Get(ctx, req.NamespacedName, vcsiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("VSphereCSIConfig resource not found")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Unable to fetch VSphereCSIConfig resource")
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := r.getOwnerCluster(ctx, vcsiConfig)
+	if cluster == nil {
+		return ctrl.Result{RequeueAfter: 20 * time.Second}, err // retry until corresponding cluster is found
+	}
+
+	return r.reconcileVSphereCSIConfig(ctx, vcsiConfig, cluster)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VSphereCSIConfigReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager,
+	options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.VSphereCSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify VSphereCSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch VSphereCSIConfig")
+			retErr = err
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileVSphereCSIConfigNormal(ctx, csiCfg, cluster); err != nil {
+		logger.Error(err, "Error reconciling VSphereCSIConfig")
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (ctrl.Result, error) {
+
+	logger := log.FromContext(ctx)
```

Comment:
suggestion: would it a good idea to pass in the logger defined at https://github.com/vmware-tanzu/tanzu-framework/pull/1942/files#diff-3410610ce615475fb7620f1c681cff9349758e47a817f2c178706e9507648413R79 ?

Reply:
logging reworked